src/root.rs

```diff
@@ -289,3 +294,66 @@ fn rm_prev_sock_path(prev_sock_path: &Path) {
         }
     }
 }
+
+// On Darwin Unix Sockets are not automatically removed, and linger,
+// with processes that try to connect to them just hanging. This makes them
+// unsuitable for our needs. Just use a file that we lock exclusively.
+#[cfg(target_os = "macos")]
+pub fn mk_lock(path: &Path) -> Result<fs::File> {
+    let lock_file = fs::File::create(path)?;
+    lock_file.lock_exclusive()?;
+    Ok(lock_file)
+}
+
+// On Linux we can use Unix Sockets as they disappear automatically,
+// which is nice.
+#[cfg(not(target_os = "macos"))]
+pub fn mk_lock(path: &Path) -> Result<UnixListener> {
+    use std::os::unix::net::UnixStream;
+
+    let socket = UnixListener::bind(path)?;
+
+    assert!(UnixStream::connect(path).is_ok());
+
+    Ok(socket)
+}
+
+#[cfg(target_os = "macos")]
+pub fn try_lock(path: &Path) -> Result<fs::File> {
+    let lock_file = fs::File::open(path)?;
+    Ok(lock_file)
+}
+
+#[cfg(not(target_os = "macos"))]
+pub fn try_lock(path: &Path) -> Result<UnixStream> {
+    let socket = UnixStream::connect(path)?;
+
+    Ok(socket)
+}
+
+#[cfg(target_os = "macos")]
+pub fn clear_lock(file: fs::File, path: &Path) -> Result<()> {
+    file.lock_exclusive()?;
+
+    // We want to unlock the file, even if we failed to remove it
+    let rm_res = fs::remove_file(path);
+
+    file.unlock()?;
+
+    // if removing failed, report it now
+    rm_res?;
+
+    Ok(())
+}
+
+#[cfg(not(target_os = "macos"))]
+pub fn clear_lock(mut s: UnixStream, _path: &Path) -> Result<()> {
+    // we are just waiting to get disconnected here
+
+    use std::io::Read as _;
+
+    // ignore, we *will* disconnect, nothing interesting about it
+    let _ = s.read(&mut [0]);
```

Comment:
@dpc maybe calling read was reason it was hanging on macos?

just `drop(s)` feels like the better implementation to me :shrug:

Reply:
It is supposed to hang. We're waiting for whoever is holding the lock to disconnect.

It was hanging on Darwin, because on Darwin when the process creating/binding Unix Socket is gone, the socket remains (unlike on Linux), and attempts to connect to it do not disconnect but just wait forever until something new binds on it.

Reply:
so clear_lock() is for waiting for lock?